Database

MollieS opened the pull request.

felipesere commented on src/main/java/pkmncore/DBManager.java:
```diff
@@ -0,0 +1,112 @@
+        try {
+            Class.forName(driver);
+        } catch (ClassNotFoundException e) {
+            throw new PokemonError("Invalid driver");
```
The exception would be more expressive it was called something along the lines of "InvalidDriverAppology" and probably took the actual driver name as an argument.

Otherwise its just a wrapper without a real purpose. It's not adding any extra information or value...

felipesere commented on src/main/java/pkmncore/DBManager.java:
```diff
@@ -0,0 +1,112 @@
+
+    private void savePokemonDetails(String name, String height, String[] abilities) throws PokemonError {
+        try {
+            Connection connection = getConnection();
```
There is a _very interesting_ repetition here and some methods down the line:
* Create a connection
* Create a statement
* **do something**
* Close the connection
* Catch a SQL exception

The only thing that changes is the **"do something"** and the error in the SQL exception.

Do you think you can come up with a clever abstraction to avoid this duplication? Show it to Rabea and have a chat with her, as I showed her a technique when she was here at Ford.

felipesere commented on src/main/java/pkmncore/DBManager.java:
```diff
@@ -0,0 +1,112 @@
+        this.driver = driver;
+    }
+
+    public Connection getConnection() throws PokemonError, SQLException {
```
Does this need to be public?

felipesere commented on src/main/java/pkmncore/DBManager.java:
```diff
@@ -0,0 +1,112 @@
+        savePokemonDetails(name, height, abilities);
+    }
+
+    public List<List<String>> getPokemon() throws PokemonError {
```
This just feels like an alias for `getAllPokemon(..)`

felipesere commented on src/main/java/pkmncore/DBManager.java:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+
+    private List<List<String>> listPokemon(Connection connection, ResultSet caughtPokemon) throws SQLException {
```
Why does this return a List<List<String>> and not a List<Pokemon>?

felipesere commented on src/main/java/pkmncore/DBManager.java:
```diff
@@ -0,0 +1,112 @@
+        return allPokemon;
+    }
+
+    private void getAbilities(List<String> pokemon, Connection connection) throws SQLException {
```
Its a little weird to read `getAbilities` but to then see that it doesn't return anything.
Using one of the inputs (List<String> pokemon) as an implicit output (the method basically says "Hey, I will change that for you!") is a bit secretive...

felipesere commented on src/main/java/pkmncore/DBManager.java:
```diff
@@ -0,0 +1,112 @@
+
+    private void listAbilities(List<String> pokemon, ResultSet abilities) throws SQLException {
+        while (abilities.next()) {
+            pokemon.add(abilities.getString("ability"));
```
Same as above...

felipesere commented on src/main/java/pkmncore/PokemonManager.java:
```diff
@@ -0,0 +1,67 @@
+        String height = pokemonDetails.remove(0);
+        String[] abilities = getAbilities(pokemonDetails);
+        Pokemon pokemon = new NamedPokemon(name, height, abilities);
+        caughtPokemon.add(pokemon);
```
Same issue as the other methods that return nothing but mutate a list.
Just have them return a fresh list of things and either re-assign it outside or just do an 
firstList.addAll(convertDetails(pokeDetails));

felipesere commented on src/main/java/pkmncore/PokemonManager.java:
```diff
@@ -0,0 +1,67 @@
+            pokemon = storage.getPokemon();
+            return pokemon;
+        } catch (PokemonError pokemonError) {
+            return pokemon;
```
This feels dodgy... An error happens and we return "an empty Pokémon"?

felipesere commented on src/main/java/pkmncore/StorageUnit.java:
```diff
@@ -0,0 +1,10 @@
+
+    void save(String name, String height, String[] abilities) throws PokemonError;
+
+    List<List<String>> getPokemon() throws PokemonError;
```
Shouldn't we be using our Pokemon abstraction here?

MollieS commented:
@felipesere I've addressed most of your comments, I think.  I haven't figured out a way to deal with the duplication of open connection/close connection.  I did try and abstract it before, but ran into issues where I was closing the connection and/or statement before I had extracted the data from the ResultSet returned.  I will talk to rabea when she is free to get some advice though so hopefully can fix that soon!
